Add Layout type and ScalarType

charliechen0401 opened the pull request:
Summary: Creating those two jit types.

Test Plan: Need suggestions on what to add unit/integration tests

Differential Revision: D18792522

smessmer commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1641,4 +1643,47 @@ struct CAFFE2_API InterfaceType : public NamedType {
+struct CAFFE2_API ScalarTypeType : public Type {
+  static ScalarTypeTypePtr create() {
+    return ScalarTypeTypePtr(
+        new ScalarTypeType()); // NOLINT(modernize-make-shared)
```
let's use `make_shared` here

Commit:
Add Layout type and ScalarType (#30694)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/30694

Creating those two jit types.

Test Plan: Need suggestions on what to add unit/integration tests

Differential Revision: D18792522

fbshipit-source-id: 995396d06da77fc93eea0ed79f048cad6b34a4c8

suo reviewed (changes requested):
Hi, changes to the fundamental JIT type system ought to have a JIT team member reviewing in the loop 😄. I have a few questions:
1. What is the purpose of this change? The PR description has no information.
2. What tests verify that the change is correct? There are no unit tests added in this PR.
Thanks!

smessmer commented:
Background: TensorOptions (i.e. ScalarType, Layout, Device, pin_memory) arguments are relevant for dispatch because tensor factories use them to dispatch to different backends. In the unboxed world, this is not an issue, because the c++ type tells you if something was a ScalarType or just an int. But in the boxed world, i.e. in a jit stack, this information is lost and ScalarType values are just stored as Ints in IValue. So the dispatcher needs to get this information from the FunctionSchema instead.

This PR introduces the ability for the FunctionSchema to have the information that the dispatcher needs here.

In my understanding, having these types map to IntType in the function_schema_parser was a hack that wasn't meant to stay anyways and this PR is fixing that hack.

smessmer commented:
The lack of test cases is my fault. @charliechen0401 asked me how to test this and I couldn't find the test folders and told him it's ok to test it together with the changes we're planning to stack on top. Can you point to the test folders?

suo commented:
Hey, here are a few open questions that we will need to answer before we can merge this:
- Backward compatibility—will existing models be able to load and run with these new changes? I think probably not, because existing models may serialize layout and scalar as ints, and LayoutType/DeviceType should probably not be implicitly convertible from ints by default. We need an upgrade path so that production models do not break.
- Forward compatibility—Will older versions of PyTorch be able to interpret the new types? This is not necessarily a dealbreaker but we need to understand what the behavior is here.
- Runtime representation—I don't see additions of Layout/Scalar IValue tags, which implies that the runtime representation is an `int`. This will lead to subtle mismatch bugs in the future, because the type of the IValue will not properly reflect the type in the type system.
- Enum-ish behavior—we have a constrained set of "correct" values the a DeviceType instance can take at runtime. How do we specify and enforce this?
- Serialization and deserialization—how do we represent Device/Scalar in the serialized format? What do they look like? If a DeviceType or ScalarType is an attribute/constant of a module, is this fact preserved across save and load?

charliechen0401 commented:
After discussing with @suo offline, we agreed on doing the following step by step:
Step 1: Introducing new type like ScalarTypeType and LayoutType but not using them in any place at this time. Also to make code looks better, introducing a new Enum base type (which is "Enum-ish behavior" in @suo 's comment above).
Step 2: Updating IValue logic, including "Runtime representation" and "Serialization and deserialization" part in the above comment.
Step 3: Using LayoutType and ScalarTypeType in the real world, and then check with other folks on Backward compatibility issue. Very likely, we will need to do something specific for upgrading path so existing models don't break after the changes. (I don't too much idea now, but we can sync with @suo and others at this point).

cc @smessmer , @suo , @zdevito

zasdfgbnm commented:
What is the current state of this? JIT is still using int to represent scalar type https://github.com/pytorch/pytorch/issues/47964. Can I change this behavior?
